### PR TITLE
NIFI-8325: Complete SNMP refactor: SNMP GET and SET processors rework, unit tests added

### PR DESCRIPTION
**Pull request by Lehel44**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

- complete SNMP refactor
- restructured with factories, builders, configuration objects
- refined error handling, validation
- unit tests by SNMP version

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java [hunk 1]
```diff
@@ -151,7 +152,24 @@ public ValidationResult validate(final String subject, final String value, final
...
+            return new ValidationResult.Builder().subject(subject).input(value).explanation(reason).valid(isValidInetAddress).build();
+        }
+
+        private boolean isValid(final String email) {
+            Matcher matcher = ipv4Pattern.matcher(email);
+            return matcher.matches();
+        }
+    };
+
+    public static final Validator PORT_VALIDATOR = createLongValidator(0, 65535, true);
```
Comment (exceptionfactory) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:172]: This change breaks existing unit tests that expect `0` to be an invalid port number.  Is there a particular reason for this change?  If so, the existing unit test should be updated, but having `0` as a port number may not work as expected in some existing components.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]

**Comment by simonbence** [comment]
Reviewing

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 2]
```diff
@@ -16,413 +16,248 @@
...
-     * further processing.
-     */
-    @Override
-    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
-        synchronized (this) {
-            this.buildTargetResource(context);
+    @OnScheduled
+    public void initSnmpClient(ProcessContext context) throws InitializationException {
+        final int version = SNMPUtils.getVersion(context.getProperty(SNMP_VERSION).getValue());
+        SNMPConfiguration configuration;
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:216]: `configuration` could be final.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java [hunk 3]
```diff
@@ -16,184 +16,188 @@
...
 })
-public class GetSNMP extends AbstractSNMPProcessor<SNMPGetter> {
+public class GetSNMP extends AbstractSNMPProcessor {
+
+    // SNMP strategies
+    public static final AllowableValue GET = new AllowableValue("GET", "GET",
+            "A manager-to-agent request to retrieve the value of a variable. A response with the current value returned.");
+    public static final AllowableValue WALK = new AllowableValue("WALK", "WALK",
+            "A manager-to-agent request to retrieve the value of multiple variables. Snmpwalk also traverses all subnodes" +
+                    "under the specified OID.");
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java:69]: Missing space: "subnodesunder"

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java [hunk 4]
```diff
@@ -16,184 +16,188 @@
...
+            performSnmpGet(context, processSession, oid);
+        } else if (SNMPStrategy.WALK == snmpStrategy) {
+            performSnmpWalk(context, processSession, oid);
+        }
+    }
+
+    private void performSnmpWalk(final ProcessContext context, final ProcessSession processSession, final String oid) {
+        try {
+            final SNMPWalkResponse response = snmpRequestHandler.walk(oid);
+            FlowFile flowFile = createFlowFileWithTreeEventProperties(response, processSession);
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java:146]: `flowFile` could be final.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java [hunk 5]
```diff
@@ -16,184 +16,188 @@
...
+            processSession.transfer(flowFile, REL_SUCCESS);
+        } catch (SNMPWalkException e) {
+            getLogger().error("Could not perform SNMP Walk: Check if SNMP agent is accessible and the specified OID exists.");
+            context.yield();
+        }
+    }
+
+    private void performSnmpGet(final ProcessContext context, final ProcessSession processSession, final String oid) {
+        final SNMPResponse response;
+        FlowFile flowFile = processSession.create();
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java:157]: `flowFile` could be final.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java [hunk 6]
```diff
@@ -16,184 +16,188 @@
...
-            .description("The OID to request")
+            .description("Each OID (object identifier) identifies a variable that can be read or set via SNMP.")
             .required(true)
-            .addValidator(SNMPUtils.SNMP_OID_VALIDATOR)
+            .addValidator(new OIDValidator())
+            .build();
+
+    public static final PropertyDescriptor SNMP_STRATEGY = new PropertyDescriptor.Builder()
+            .name("snmp-strategy")
+            .displayName("SNMP Strategy (GET / WALK)")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/GetSNMP.java:82]: The options in the list show the possible values (GET and WALK). I think it can be removed from the displayName.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html [hunk 7]
```diff
@@ -54,16 +54,18 @@ <h2>Configuration Details</h2>
...
     <li><b>Authentication Password</b> - [REQUIRED if SNMPv3 and authNoPriv or authPriv] the authentication password to use.</li>
-    <li><b>Private Protocol</b> - [REQUIRED if SNMPv3 and authPriv] the private protocol to use.</li>
-    <li><b>Private Password</b> - [REQUIRED if SNMPv3 and authPriv] the private password to use.</li>
-    <li><b>Retries</b> - [REQUIRED] Sets the number of retries to be performed before a request is timed out. (default is 0)</li>
-    <li><b>Timeout</b> - [REQUIRED] Sets  timeout in milliseconds before a confirmed request is resent or timed out. (default is 5000)</li>
+    <li><b>Privacy Protocol</b> - [REQUIRED if SNMPv3 and authPriv] the privacy protocol to use.</li>
+    <li><b>Privacy Password</b> - [REQUIRED if SNMPv3 and authPriv] the privacy password to use.</li>
+    <li><b>Retries</b> - [NOT REQUIRED] Sets the number of retries to be performed before a request is timed out. (default is 0)</li>
+    <li><b>Timeout</b> - [NOT REQUIRED] Sets  timeout in milliseconds before a confirmed request is resent or timed out. (default is 1000)</li>
+    <li><b>Textual OID</b> - [NOT REQUIRED] Optionally added to the flow file</li>
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html:68]: Are these the same properties as the PropertyDescriptors in the processor?
If yes, the same names should be used as on the UI (eg. Password vs Passphrase).
I feel a bit overkill to maintain this list (unless it contains additional info, but I don't really see it).

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java [hunk 8]
```diff
@@ -151,7 +152,24 @@ public ValidationResult validate(final String subject, final String value, final
         }
     };
 
-    public static final Validator PORT_VALIDATOR = createLongValidator(1, 65535, true);
+    public static final Validator NETWORK_ADDRESS_VALIDATOR = new Validator() {
+        private final Pattern ipv4Pattern = Pattern.compile("^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\\.(?!$)|$)){4}$");
```
Comment (turcsanyip) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:156]: `ipv4Pattern` could be static.

**Review by turcsanyip — CHANGES_REQUESTED** [review]
@Lehel44 Thanks for refactoring these old processors!
I read through the code and added some comments regarding UI stuff (display names, etc) and minor code issues.
I'll test the processors later.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 9]
```diff
@@ -16,413 +16,248 @@
...
+    // SNMPv3 authentication protocols
+    public static final AllowableValue MD5 = new AllowableValue("MD5", "MD5 based authentication",
+            "Provides authentication based on the HMAC-MD5 algorithm.");
+    public static final AllowableValue SHA = new AllowableValue("SHA", "SHA based authentication",
+            "Provides authentication based on the HMAC-SHA algorithm.");
+    public static final AllowableValue NO_AUTHENTICATION = new AllowableValue("", "None",
+            "Sends SNMP requests without authentication.");
+
+    // SNMPv3 encryption
+    public static final AllowableValue DES = new AllowableValue("DES", "DES",
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:72]: DES is not considered secure due to the ease of brute-force and other attacks.  Is there a compelling reason to provide it as an allowable option?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 10]
```diff
@@ -16,413 +16,248 @@
...
+            "Symmetric-key algorithm for the encryption of digital data.");
+    public static final AllowableValue DES3 = new AllowableValue("3DES", "3DES",
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:74]: Although 3DES provides a much greater level of security than DES, it is no longer recommend for use, is there a reason it should be allowed?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 11]
```diff
@@ -16,413 +16,248 @@
...
+                    "SNMPv2c and SNMPv3, feature improvements in performance, flexibility and security.")
             .required(true)
-            .allowableValues("SNMPv1", "SNMPv2c", "SNMPv3")
-            .defaultValue("SNMPv1")
+            .allowableValues(SNMP_V1, SNMP_V2C, SNMP_V3)
+            .defaultValue(SNMP_V1.getValue())
             .build();
 
-    /** property to define SNMP community to use */
     public static final PropertyDescriptor SNMP_COMMUNITY = new PropertyDescriptor.Builder()
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:116]: SNMP Community should be marked as a sensitive property since it serves authentication purposes.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/resources/log4j.properties [hunk 12]
```diff
@@ -13,8 +13,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 log4j.rootLogger=INFO, CONSOLE
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/resources/log4j.properties:15]: Is this file necessary?  Log4j calls should be routed to SLF4J and this configuration file should be removed.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java (hunk 9 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:72]: It depends on the user settings. If the agents' user security model (USM) is configured with DES, I think we should allow this option. I'm thinking of extending the option with a warning about DES vulnerability. What do you think?

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:72]: That seems reasonable, including a warning in the explanation sounds good.

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java (hunk 1 above)
Comment (Lehel44) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:172]: I think 0 shouldn't be invalid, although we could make more validators based on strictness. According to [wikipedia](https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers)
- 0: In programming APIs (not in communication between hosts), requests a system-allocated (dynamic). java.net.InetSocketAddress is designed this way (A valid port value is between 0 and 65535. A port number of zero will let the system pick up an ephemeral port in a bind operation.) and according the this standard we can assume other API's follow it too.
- 0-1023: well known ports
- 1024-49151: Registered ports by IANA
- 49152-65535: Ephemeral ports typically used by TCP and UDP connections

I'm thinking about a few options.

- Creating different PORT validators for well known, registered and ephemeral ports.
- Including 0 or not with any of these.

Comment (exceptionfactory) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:172]: Thanks for the reference, it seems like `0` should be acceptable in some cases where a random port number could be useful.  Some existing components rely on an explicit port number, so I think the best approach is to leave the existing `PORT_VALIDATOR` as it is, but create a new port validator that allows `0` to indicate preference for a dynamic port number.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 13]
```diff
@@ -16,413 +16,248 @@
...
+    public static final AllowableValue AES256 = new AllowableValue("AES256", "AES256", AES_DESCRIPTION);
+    public static final AllowableValue NO_ENCRYPTION = new AllowableValue("", "None",
+            "Sends SNMP requests without encryption.");
+
+
+    public static final PropertyDescriptor AGENT_HOST = new PropertyDescriptor.Builder()
             .name("snmp-hostname")
-            .displayName("Host Name")
-            .description("Network address of SNMP Agent (e.g., localhost)")
+            .displayName("SNMP Agent Hostname")
```
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:89]: Localhost is allowed also, if we support hostnames we don't need a regex validator as it can be any string. Can we use something like InetAddress.getByName(String)? I can change the validator. If it successfully performs a DNS lookup, the hostname is valid, otherwise it isn't.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/resources/log4j.properties (hunk 12 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/test/resources/log4j.properties:15]: It's indeed not. Thanks, I'll remove it.

**Comment by Lehel44** [comment]
I changed the SNMPUtils::getAuth and SNMPUtils::getPriv because according to the snmp4j documentation the auth and priv protocols can be null. The map had already contained the empty string-null key value pair but the optionals threw exception instead of returning the null value.

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java (hunk 1 above)
Comment (Lehel44) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:172]: It seems like every property needs a specific port. For now we don't need the extended validator. I'll change this back to original.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:89]: As we discussed, domain name and ip validation together is complicated. It is up to the user to enter the correct domain or IP address, I'll replace this with a non empty validator and remove the network address validator from StandarValidators.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html (hunk 7 above)
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html:68]: I feel likewise, I'll remove the list of properties.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/factory/V3SNMPFactory.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        final String username = configuration.getSecurityName();
+        final String authProtocol = configuration.getAuthProtocol();
+        final String authPassword = configuration.getAuthPassword();
+        final String privacyProtocol = configuration.getPrivacyProtocol();
+        final String privacyPassword = configuration.getPrivacyPassword();
+        final OctetString authPasswordOctet = authPassword != null ? new OctetString(authPassword) : null;
+        final OctetString privacyPasswordOctet = privacyPassword != null ? new OctetString(privacyPassword) : null;
+
+        // Add user information.
+        snmp.getUSM().addUser(
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/factory/V3SNMPFactory.java:59]: It seems SNMP4J library stores the authenticated users in static context in the JVM.
Eg. when I configure a processor instance with valid user credentials and start it up, then other processors using the same username but invalid credentials can also authenticate to the agent and run without error.
To avoid this interference, `@RequiresInstanceClassLoading` may need to be used.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 15]
```diff
@@ -16,413 +16,249 @@
...
-            .name("snmp-private-protocol")
-            .displayName("SNMP Private Protocol")
-            .description("SNMP Private Protocol to use")
+    public static final PropertyDescriptor SNMP_AUTH_PROTOCOL = new PropertyDescriptor.Builder()
+            .name("snmp-authentication-protocol")
+            .displayName("SNMP Authentication Protocol")
+            .description("Hash based auhentication protocols for secure authentication.")
             .required(true)
-            .allowableValues("DES", "3DES", "AES128", "AES192", "AES256", "")
+            .allowableValues(MD5, SHA, NO_AUTHENTICATION)
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:178]: `None` option seems to be obsolete because this property depends on Security Level and when the property is visible, it is mandatory.

Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:178]: SHA-2 based protocols are also available in SNMP4J (eg. `org.snmp4j.security.AuthHMAC192SHA256`) and we should support them too.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 16]
```diff
@@ -16,413 +16,249 @@
...
+            .dependsOn(SNMP_VERSION, SNMP_V3)
             .build();
 
-    /** property to define SNMP security name to use */
     public static final PropertyDescriptor SNMP_SECURITY_NAME = new PropertyDescriptor.Builder()
             .name("snmp-security-name")
-            .displayName("SNMP Security name / user name")
-            .description("Security name used for SNMP exchanges")
-            .required(false)
+            .displayName("SNMP Security Name / Username for SNMP v3 Authentication")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:146]: This property name is too long and does not fit into the cell on the UI. I would use `SNMP Security Name` only and add the explanation in the description.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 17]
```diff
@@ -16,413 +16,249 @@
...
-            .name("snmp-authentication-protocol")
-            .displayName("SNMP Authentication Protocol")
-            .description("SNMP Authentication Protocol to use")
+    public static final PropertyDescriptor SNMP_PRIVACY_PROTOCOL = new PropertyDescriptor.Builder()
+            .name("snmp-private-protocol")
+            .displayName("SNMP Privacy Protocol")
+            .description("Privacy allows for encryption of SNMP v3 messages to ensure confidentiality of data.")
             .required(true)
-            .allowableValues("MD5", "SHA", "")
+            .allowableValues(DES, DES3, AES128, AES192, AES256, NO_ENCRYPTION)
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:158]: `None` option seems to be obsolete because this property depends on Security Level and when the property is visible, it is mandatory.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 18]
```diff
@@ -16,413 +16,249 @@
...
-    /** property to define SNMP authentication password to use */
-    public static final PropertyDescriptor SNMP_AUTH_PASSWORD = new PropertyDescriptor.Builder()
-            .name("snmp-authentication-passphrase")
-            .displayName("SNMP Authentication pass phrase")
-            .description("Pass phrase used for SNMP authentication protocol")
-            .required(false)
+    public static final PropertyDescriptor SNMP_PRIVACY_PASSWORD = new PropertyDescriptor.Builder()
+            .name("snmp-private-protocol-passphrase")
+            .displayName("SNMP Privacy Protocol Passphrase")
+            .description("Pass phrase used for SNMP privacy protocol")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:166]: Passphrase is written as a single word in other places nearby.
(Minor: there should be a period at the end of the sentence.)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 19]
```diff
@@ -16,413 +16,249 @@
...
 
-    /** property to define SNMP authentication password to use */
-    public static final PropertyDescriptor SNMP_AUTH_PASSWORD = new PropertyDescriptor.Builder()
-            .name("snmp-authentication-passphrase")
-            .displayName("SNMP Authentication pass phrase")
-            .description("Pass phrase used for SNMP authentication protocol")
-            .required(false)
+    public static final PropertyDescriptor SNMP_PRIVACY_PASSWORD = new PropertyDescriptor.Builder()
+            .name("snmp-private-protocol-passphrase")
+            .displayName("SNMP Privacy Protocol Passphrase")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java:165]: I would use `SNMP Privacy Passphrase`.
(similar to `SNMP Authentication Passphrase`)

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SetSNMP.java [hunk 20]
```diff
@@ -34,191 +25,101 @@
...
-                    context.yield();
-                }
-            } else {
+            try {
+                final SNMPResponse response = snmpRequestHandler.set(flowFile);
+                processResponse(processSession, flowFile, response);
+            } catch (SNMPException | IOException e) {
                 processSession.transfer(processSession.penalize(flowFile), REL_FAILURE);
-                this.getLogger().warn("No attributes found in the FlowFile to perform SNMP Set");
+                getLogger().error("Could not perform SNMP Set: Request timed out or parameters are incorrect.");
```
[the remark this answers is not in the document]
Comment (simonbence) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SetSNMP.java:91]: Also based on some testing: the processor provides only a limited amount of information when the set is failed. It would be good to have more information on the interface. Also: I think timeout might be handled differently. It is possible "rolling back" the flow file and then yielding would be better, as possibly there is no issue with the input but we have a temporary outage in the other end

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/dto/SNMPResponse.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+public class SNMPResponse {
+
+    private final Map<String, String> attributes;
+    private final List<SNMPValue> variableBindings;
+    private final int errorStatus;
+    private final String errorStatusText;
+    private final String targetAddress;
+
+    public SNMPResponse(final PDU responsePdu, final Target target) {
+        attributes = SNMPUtils.getPduAttributeMap(responsePdu);
```
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/dto/SNMPResponse.java:36]: I extract information on the fly when SNMP Get is being called here: [line](https://github.com/apache/nifi/blob/e121973906f72899354011fdeeb48538fba99a94/nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/operations/StandardSNMPRequestHandler.java#L72)

We extract the information only once, why do you think we should store it?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/factory/AbstractSNMPFactory.java [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+
+public abstract class AbstractSNMPFactory {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractSNMPFactory.class);
+
+    protected AbstractSNMPFactory() {
+        // hide implicit constructor
+    }
+
+    protected static Snmp createSimpleSnmpClient() {
```
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/factory/AbstractSNMPFactory.java:43]: I removed the simple from the signature. @turcsanyip recommended using create instead of get because its created when called.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html [hunk 23]
```diff
@@ -44,27 +44,5 @@ <h2>SNMP Properties</h2>
     In case of a single SNMP Get request, the following is the list of available standard SNMP properties which may come with the PDU:
     <i>("snmp$errorIndex", "snmp$errorStatus", "snmp$errorStatusText", "snmp$nonRepeaters", "snmp$requestID", "snmp$type")</i>
 </p>
-<h2>Configuration Details</h2>
```
[the remark this answers is not in the document]
Comment (Lehel44) [reply nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/resources/docs/org.apache.nifi.snmp.processors.GetSNMP/additionalDetails.html:47]: All the information can already be found here in the property descriptions. I think it's unnecessary to maintain the same list here.

**Review by turcsanyip — APPROVED** [review]
+1 Merging to main.
Tested with SNMP agent simulator using different settings (v1, v2c, v3, auth/priv settings, error cases).

@Lehel44 Thanks for your contribution.
@simonbence, @exceptionfactory Thanks for the reviews.
